pkg/volume/gce_pd/gce_util.go

```diff
@@ -122,6 +122,167 @@ func (util *GCEDiskUtil) DetachDisk(pd *gcePersistentDisk) error {
 	if err := gce.(*gce_cloud.GCECloud).DetachDisk(pd.pdName); err != nil {
 		return err
 	}
+
+	// Verify disk detached, retry if needed.
+	go verifyDiskDetachedRetryIfNeeded(pd, gce)
+	return nil
+}
+
+// Verifys the disk device to be created has been succesffully attached, and retries if it fails.
+func verifyDiskAttachedRetryIfNeeded(pd *gcePersistentDisk, sdBeforeSet util.StringSet, gce cloudprovider.Interface) (string, error) {
+	devicePaths := getDiskByIdPaths(pd)
+	for numRetries := 0; numRetries < 10; numRetries++ {
+		for numChecks := 0; numChecks < 10; numChecks++ {
+			if err := udevadmChangeToNewDrives(sdBeforeSet); err != nil {
+				// udevadm errors should not block disk attachment, log and continue
+				glog.Errorf("%v", err)
+			}
+
+			for _, path := range devicePaths {
+				if pathExists, err := pathExists(path); err != nil {
+					return "", err
+				} else if pathExists {
+					// A device path has succesfully been created for the PD
+					glog.V(5).Infof("Succesfully attached GCE PD %q.", pd.pdName)
+					return path, nil
+				}
+			}
+
+			// Sleep then check again
+			glog.V(5).Infof("Waiting for GCE PD %q to attach.", pd.pdName)
+			time.Sleep(time.Second)
+		}
+
+		// Try attaching the disk again
+		glog.Warningf("Timed out waiting for GCE PD %q to attach. Retrying attach.", pd.pdName)
+		if err := gce.(*gce_cloud.GCECloud).AttachDisk(pd.pdName, pd.readOnly); err != nil {
+			return "", err
+		}
+	}
+
+	return "", fmt.Errorf("Could not attach GCE PD %q. Timeout waiting for mount paths to be created.", pd.pdName)
+}
+
+// Veify the specified persistent disk device has been succesfully detached, and retries if it fails.
+func verifyDiskDetachedRetryIfNeeded(pd *gcePersistentDisk, gce cloudprovider.Interface) {
+	ch := detachChanManager.CreateAndAddChan(pd.pdName, 0 /* bufferSize */)
+	defer detachChanManager.CloseAndRemoveChan(pd.pdName)
+	devicePaths := getDiskByIdPaths(pd)
+	for numRetries := 0; numRetries < 10; numRetries++ {
+		for numChecks := 0; numChecks < 10; numChecks++ {
+			select {
+			case <-ch:
+				glog.Warningf("Terminating GCE PD %q detach verification. Another attach/detach call was made for this PD.", pd.pdName)
+				return
+			default:
+				allPathsRemoved := true
+				for _, path := range devicePaths {
+					if err := udevadmChangeToDrive(path); err != nil {
+						// udevadm errors should not block disk detachment, log and continue
+						glog.Errorf("%v", err)
+					}
+					if exists, err := pathExists(path); err != nil {
+						glog.Errorf("Error check path: %v", err)
+						return
```

Comment:
Reiterating above, since this method is executed as a go routine, it doesn't make sense for it to return anything. From the [golang spec](http://golang.org/ref/spec#Go_statements): `When the function terminates, its goroutine also terminates. If the function has any return values, they are discarded when the function completes.`